## Stage- B quiz solutions

### Importing all important libraries

In [43]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler


### Question 1:
The percent of the total variation of the dependent variable Y explained by the set of independent variables X is measured by?

#### Answer 1

* Multicollinearity exists when the predictors are correlated with each other.
* Correlation coefficient tells us about the linear relationship that exists between two variables. It ranges between [-1.1]
* Coefficient of determination, that is also called as R^2 tells how much varaiability of the dependent variable is being explained by the set of our independent variables. Hence it's correct answer.
* coefficient of skewness measures the lack of symmetry present in the distribution of the data.

### Hence coefficient of determination is correct option.

### Question 2
How do you define a Residual?



#### Answer 2

Residual is the difference between the actual value and the predicted value.
Hence,  
### $e = y - y{hat}$ is correct option

### Question 3
The straight line graph of the equation Y = a + BX, the slope is horizontal if?

#### Answer 3

when b = 0, y = a i.e constant.

### Hence b = 0 is correct


#### Question 4:
Which of the one is true about Heteroskedasticity?



#### Answer 4

Homoscedasticity refers to the constant variance of error term. Contrary to it, we have Heteroskedasticity which means Linear regression with variations in error term.

### Therefore, Linear Regression with varying error terms is correct option

#### Question 5

Generally, which of the following method(s) is used for predicting continuous dependent variables?

1. Linear Regression

2. Logistic Regression

#### Answer 5
Logistic regression is used for classification problems i.e when our target variable has discrete labels where as Linear regression deals with regression problems in which target or dependent variable is of continuous type.

### Hence,  1 only is correct option.

#### Question 6
From the following options below, which of these is/are true about “Ridge” or “Lasso” regression methods in case of feature selection?

#### Answer 6

In LASSO, we use feature selection technique. LASSO helps to reduce overfitting.

### Hence, Lasso regression uses subset selection of features is correct option.

#### Question 7
Which of the following sentences is/are true about outliers in Linear Regression:

#### Answer 7
We know Linear regression is sensitive to outliers because we know 
slope will be changed and will not give better line that we could have without outliers. Also residual will be large in this case,

### Thus, Linear regression is sensitive to outliers

#### Question 8
Which of the following metrics can be used for evaluating regression models?



#### Answer 8
All of these metrics can be used for evaluation of regression models.

### Hence 1,2,3 and 4 is correct option.

#### Question 9
A best fit line relating X and Y has a R-Squared value of 0.75. How do I interpret this information?

#### Answer 9
Coefficient of determination, that is also called as R^2 tells how much varaiability of the dependent variable is being explained by the set of our independent variables.

### Thus, 75% of the variance in Y is explained by X is correct option.

#### Question 10
Which of the following measures is optimal for comparing the goodness of the fit of competing regression models involving the same dependent variable?

#### Answer 10
#### R-square is correct option.

#### Question 11
The Lasso can be interpreted as least-squares linear regression where:



#### Answer 11
LASSO deals with L1- norm hence weights are regularised with L1 norm.

### Therefore, Weights are regularized with the L1 norm is correct option.

#### Question 12
From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?


#### Answer 12

In [44]:
df = pd.read_csv('energydata_complete (1).csv')

In [45]:
df.head(2)

date  Appliances  lights     T1       RH_1    T2     RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.7900   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.7225   

      T3   RH_3    T4  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  19.79  44.73  19.0  ...  17.033333  45.53  6.600000        733.5    92.0   
1  19.79  44.79  19.0  ...  17.066667  45.56  6.483333        733.6    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  

[2 rows x 29 columns]

In [46]:
data = df[['T2','T6']]
target = df[['T6']]
feature = data.drop(['T6'],axis = 1)
X_train, X_test, y_train, y_test = train_test_split(feature,target,
test_size=0.3, random_state=42)
linear_model = LinearRegression()
#fit the model to the training dataset
linear_model.fit(X_train, y_train)
#obtain predictions on test dataset
predictions = linear_model.predict(X_test)
r_2 = r2_score(y_test,predictions)
round(r_2,2)

0.64

### Hence, 0.64 is correct option

#### Question 13
Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. Answer the following questions:

What is the Mean Absolute Error (in two decimal places)?

#### Answer 13


In [47]:
df.drop(columns = ['date','lights'],inplace = True)
scaler = MinMaxScaler()
normalised_df = pd.DataFrame(scaler.fit_transform(df),columns = df.columns)
feature_df = normalised_df.drop(columns = ['Appliances'])
target = normalised_df['Appliances']
X_train, X_test, y_train, y_test = train_test_split(feature_df,target,
test_size=0.3, random_state=42)

In [48]:
X_train.shape, X_test.shape

((13814, 26), (5921, 26))

In [49]:
y_train.shape, y_test.shape

((13814,), (5921,))

In [50]:
linear_model = LinearRegression()
#fit the model to the training dataset
linear_model.fit(X_train, y_train)
predicted_values = linear_model.predict(X_test)
#MAE
mae = mean_absolute_error(y_test, predicted_values)
round(mae, 3)

0.05

### Hence 0.05 is correct option.

#### Question 14
What is the Residual Sum of Squares (in two decimal places)?



#### Answer 14

In [51]:
rss = np.sum(np.square(y_test -predicted_values ))
round(rss,2)

45.35

### Hence 45.35 is correct option

#### Question 15
What is the Root Mean Squared Error (in three decimal places)?



#### Answer 15


In [52]:
rmse = np.sqrt(mean_squared_error(y_test,predicted_values))
round(rmse,3)

0.088

### Hence 0.088 is correct option.

#### Ques 16:
What is the Coefficient of Determination (in two decimal places)?


#### Answer 16

In [53]:
r2_score = r2_score(y_test,predicted_values)
round(r2_score,2)

0.15

### Hence 0.15 is correct option.

#### Question 17
Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

#### Answer 17

In [54]:
def weights_df(model,feature,col_name):
    weight_dataframe = pd.Series(model.coef_ , feature.columns).sort_values().reset_index()
    weight_dataframe.columns = ['Features',col_name]
    return weight_dataframe

In [55]:
linear_model_weights = weights_df(linear_model, X_train, 'Linear_Model_Weight')

In [56]:
linear_model_weights

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

Clearly most negative value is of RH_2 and highest value is of RH_1.

### RH_2, RH_1 is correct option.

#### Ques 18

Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?
#### Answer 18

In [57]:
rig_reg = Ridge(alpha = 0.4)
rig_reg.fit(X_train,y_train)
y_pred = rig_reg.predict(X_test)
rmse = np.sqrt(mean_squared_error(y_test,y_pred))
round(rmse,3)

0.088

No change 
### Hence, No is correct option.

#### Ques 19
Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?
#### Answer 19

In [58]:
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)
y_pred1 = lasso_reg.predict(X_test)
lasso_weights_df = weights_df(lasso_reg, X_train, 'Lasso_weight')

In [59]:
lasso_weights_df

Features  Lasso_weight
0        RH_out     -0.049557
1          RH_8     -0.000110
2            T1      0.000000
3     Tdewpoint      0.000000
4    Visibility      0.000000
5   Press_mm_hg     -0.000000
6         T_out      0.000000
7          RH_9     -0.000000
8            T9     -0.000000
9            T8      0.000000
10         RH_7     -0.000000
11          rv1     -0.000000
12           T7     -0.000000
13           T6      0.000000
14         RH_5      0.000000
15           T5     -0.000000
16         RH_4      0.000000
17           T4     -0.000000
18         RH_3      0.000000
19           T3      0.000000
20         RH_2     -0.000000
21           T2      0.000000
22         RH_6     -0.000000
23          rv2     -0.000000
24    Windspeed      0.002912
25         RH_1      0.017880

#### There are 4 non-zero values hence, 4 is correct option

#### Question 20
What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)

#### Answer 20

In [60]:
rmse = np.sqrt(mean_squared_error(y_test,y_pred1))
round(rmse,3)

0.094

### Hence, 0.094 is correct option.



## Thank you